### DISCUSSION TITLE
[Question:] What movements feel best to you?

### DISCUSSION THREAD

**Discussion by Glitchy-Tozier** [8VIM/8VIM#140] [Q&A]
### The importance of fluidity VS finger-travel-distance, in your opinion?

This question relates to https://github.com/flide/8VIM/discussions/138
In that program, I'm not quite sure, how to weigh fluidity and finger-travel-distance. It is needed to calculate the best layout for each language (including English).

Quick reminder: THIS is fluidity
![8vim_q1](https://user-images.githubusercontent.com/59611881/102071583-fa962c80-3e00-11eb-935b-97e1eac03346.jpg)
and THIS is finger-travel-distance
![8vim_q2](https://user-images.githubusercontent.com/59611881/102071587-fb2ec300-3e00-11eb-98ae-2b56f776358d.jpg)

## So here's my four-part-question:

**1.** How much experience do you have with 8VIM / 8pen?

**2.** If you had 10 "**importance-points**", how many would you give to **fluidity** and how many would you give to **finger-travel-distance**?

**3.** On a scale of 1 (worst) to 10 (best), how would you rate those 4 movements? (The points don't have to add up to 10.)
![8vim_q2](https://user-images.githubusercontent.com/59611881/102071587-fb2ec300-3e00-11eb-98ae-2b56f776358d.jpg)

**4.** After each of these two Movements, ...
![20201214_125557](https://user-images.githubusercontent.com/59611881/102078868-f1f72380-3e0b-11eb-885c-c2bfcbc2e451.jpg)
... how fluent is a 1-layer movement into each of the following positions?
```
   \  3           4  /   
  2  \             /  5  
       \         /       
         \     /         
           \ /           
           / \           
         /     \         
       /         \       
  1  /             \  6  
   /  0           7  \   
```
That's what I would like to know from some experienced users!

**Comment by Glitchy-Tozier** [comment]
#### My personal answer:

1. **Experience with 8VIM / 8pen:** Almost none. Still waiting for that German layout.

2. **"Importance-points":**
```
6 for good flow
4 for short distance
```
Reasoning: For layer 1&2, Flow is more important. For layer 3&4, distance gets REALLY annoying, more so than bad flow.
Since layer 1 & layer 2 are more common than layer 3 & layer 4 though, this is my answer. (instead of 5:5)

3. **How comfortable is each distance?**
```
Quarter-circle: 10/10
Half-circle: 8/10
Three-quarter-circle: 3/10
Full circle: 1/10
```
4. **Fluidity of movements**
```
Quarter-circle and...
0 = 10
1 = 1
2 = 3
3 = 7
4 = 4
5 = 9
6 = 7
7 = 7

Half-circle and...
0 = 10
1 = 5
2 = 10
3 = 0
4 = 2
5 = 7
6 = 1
7 = 10
```

**Comment by flide** [comment]
Okay, following is going to be a long post and an out pour of all my thoughts on the 'flow' problem which I have been mulling over for last 2 weeks.

## Context

1. _flow_ can be thought of in the x pad only in terms of *2* characters that need to be typed after one another.
2. _flow_ is essentially a transition from 1st character to the second character
3. since the _flow_ is a transition from 1st to 2nd Character, *it* takes place only _*inside the circle*_

## Mental Picture

Given the context there are only 3 kinds of motions possible inside the circle to transition from 1st character to the 2nd character. 
To be able to explain these three motions, please consider the circle as a *square room* with 4 doors, one at the center of each wall.
When you are completing the 1st character, you _enter_ the room from one of the doors. Let's name the door - D1.

So now, as you enter the room, the transition starts and you are going to _flow_ into the second character as you _exit_ from one of the doors.

Let's name the other doors - 
- D1 - door we entered from 
- D2 - door to the right of D1
- D3 - door opposite to D1
- D4 - door to the left of D1

Since there exist 4 doors, you have 4 ways to exit and hence 4 (but really 3) _flow_ possibilities. 
- You exit through D3 by running straight through. This has to be the best in terms of fluidity. It's the least direction change, hence least effort to change momentum. In my opinion - Best flow possible.
- You exit though D2, or D4. This is *okay* in terms of _flow_. you can arc graciously from D1 to D2 or D3. I hope, you can see how these 2 exits represent exact same when we are taking about _flow_.
- You exit through D1. This is the worst of all the options. You either need to stop and do a 180 degree turn or if you want to maintain fluidity, you will have to circle the entire room and **not slip** through any other door and exit through D1. This is the worst for the _flow_

## Calculations to figure out *Fluidity* of a given layout
So, we have the following information
- *Fluidity* of the board will depend on the all possible combinations of bi-grams on the board. It should be the sum of the _flow_ of all possible bi-grams each multiplied by the frequency of it's occurrence in the language. Since there are 32 possible characters displayed at a time, it is going to be the sum of (32*32 = 1024) terms.
`Fluidity = Summation(Flow(i,j) * Frequency(i,j))`
Where, Flow(i,j) is the calculated flow from character i to j
and, Frequency(i,j) is the frequency of bi-gram 'ij' in the language. also frequency(i,j) is not equal to frequency(j,i)
and, Summation() represents the addition of all 1024 terms.
- Now the question boils down to - How are we going to *refine* _flow_ into a mathematical number. We essentially have 2 choices : 
   * scale the flows from low to high - (1, 2, 3)
   * scale the flows from negative to positive - (-1, 0, 1). This is going to destroy the summation term for most of the bi-grams, so this is really not a choice.

**Reply by Glitchy-Tozier** [reply]
First of all, thank you for taking the time to write all of this! 
Let me respond to a few points:

> 1. flow can be thought of in the x pad only in terms of 2 characters that need to be typed after one another.

100% agree!

> 2. flow is essentially a transition from 1st character to the second character

Yes

> 3. since the flow is a transition from 1st to 2nd Character, it takes place only inside the circle

No, unfortunately. Let me explain.
You say this…
> • D1 - door we entered from
 • D2 - door to the right of D1
 • D3 - door opposite to D1
 • D4 - door to the left of D1

… and that _is_ mostly true. However, there are exceptions. Take a look at this:
![8vim_flno](https://user-images.githubusercontent.com/59611881/102370629-95d40100-3fbd-11eb-8e30-4e1c81693e10.jpg)
You agree that those two have a different flow, right? (not _extremely_ different, but different enough.
> Since there exist 4 doors, you have 4 ways to exit and hence 4 (but really 3) flow possibilities.

This is a nice way of thinking about it. I think we should try to be _even more_ precise: **8 Options**.
This way, we could differentiate between the two movements in the picture.


________________________

>  `Fluidity = Summation(Flow(i,j) * Frequency(i,j))`

That's EXACTLY, what I'm doing in my script!!  :)

I'd like to get your opinion on the four questions in the first post.
> The importance of fluidity VS finger-travel-distance, in your opinion?
Could you answer those questions?
____________
> Now the question boils down to - How are we going to refine flow into a mathematical number. We essentially have 2 choices:
    scale the flows from low to high - (1, 2, 3)
    scale the flows from negative to positive - (-1, 0, 1). This is going to destroy the summation term for most of the bi-grams, so this is really not a choice.

I chose the first, as you can see, if you read the top post and answer the questions.  ;)
______________________
Btw, my scipt is almost finished!
(it's also  very long. https://github.com/Glitchy-Tozier/8vim_keyboard_layout_calculator)
Now I just need a few people to answer my **four questions**.

**Reply by flide** [reply]
I don't fully agree with the distiction of the 2 movements you have presented. I will let others weigh in on the discussion, before I respond.

**Reply by Glitchy-Tozier** [reply]
If you don't fully agree,  please show me so by answering the four questions in my **top post.**
If something about the questions is unclear, make sure to ask!

**Reply by Glitchy-Tozier** [reply]
Another question, I'm curious:

Does the first movement seem worse to you?
To me, they seem equally good. Maybe the first movement feels even _slightly_ better, because it stays in one spot and doesn't travel halfway across the screen.
![8vim_flno2](https://user-images.githubusercontent.com/59611881/102457850-b0e95400-4043-11eb-8877-32c1f2310ecf.jpg)
_______________________
And how about this?
Same 'Door', but the first movement seems _way_ worse to me.
![8vim_flno3](https://user-images.githubusercontent.com/59611881/102458392-77651880-4044-11eb-9abd-c9e9fae7f5a7.jpg)

**Reply by kjoetom** [reply]
I also think like @Glitchy-Tozier that there is a difference here. And in my opinion, this difference can be determined by two factors. 
First, whether there is a change between "clockwise" and "counterclockwise" or not. 
And on the other hand, whether and how far the center of gravity of the movement shifts between the first and the second letter.

But I have no idea how to capture and weight this mathematically.

**Comment by flide** [comment]
#### My Answer to the Original Question

1. **Experience with 8VIM / 8pen:** Using it for almost *9 years*. I *hate* having to type on the regular keyboards now. It really frustrates me.
2. **"Importance-points":**
```
7 for good flow
3 for short distance
```
Reasoning: In my mind the _long-swipe-distance_ is only relatively long. on a 2x2 inch area it really does not matter that much.
3. **How comfortable is each distance?**
```
Quarter-circle: 10/10
Half-circle: 9/10
Three-quarter-circle: 8/10
Full circle: 5/10
```
4. **Fluidity of movements**
I don't understand the fourth question completely... Not going to provide a misguided answer.

**Reply by Glitchy-Tozier** [reply]
Thank you, this is EXACTLY what I needed from you! It's really well formatted too!

Regarding the 4th question: No problem, if you tell me how fluent each of the doors is (again, on a scale of 1 to 10), I think I know your answer.
>• D1 - door we entered from
• D2 - door to the right of D1
• D3 - door opposite to D1
• D4 - door to the left of D1

**Reply by flide** [reply]
D1 to D3 -> Fluidity 10/10
D1 to D2/D4 -> Fluidity 8/10
D1 to D1 -> Fluidity 4/10 (I can actually make it work, but that requires a larger circle, and I am trying to get to the smallest circle I possibly can to reduce the loop lengths, but that is the way forward for me for diminishing returns)

**Comment by ghost** [comment]
Hi there,

Sorry for my late reply, I hope I’m not too late.

Here are my answers to your questions:

### Experience with 8Vim: 
Close to none. Just toyed around a little but I got stuck with the German diacritics.

### "Importance-points":
I would differ between the layers since distance becomes more important at higher layers:
* layer 1 & 2: flow/distance → 8/2
* layer 3: flow/distance → 4/6 
* layer 4: flow/distance → 2/8

### How comfortable is each distance?
* layer 1: 10/10
* layer 2: 8/10
* layer 3: 5/10
* layer 4: 2/10
### Fluidity of movements
* layer 1 to 0: 10/10
* layer 1 to 1: 2/10
* layer 1 to 2: 2/10
* layer 1 to 3: 7/10
* layer 1 to 4: 5/10
* layer 1 to 5: 10/10
* layer 1 to 6: 10/10
* layer 1 to 7: 8/10

* layer 2 to 0: 10/10
* layer 2 to 1: 7/10
* layer 2 to 2: 8/10
* layer 2 to 3: 2/10
* layer 2 to 4: 2/10
* layer 2 to 5: 5/10
* layer 2 to 6: 7/10
* layer 2 to 7: 10/10

I hope, this helps you.

Cheers,
Tristan

**Reply by Glitchy-Tozier** [reply]
Hey, thank you for the answer!  :)
Hopefully I'll be done with my script soon, I also want to finally be able to write German texts!

By the way, if you're feeling motivated (and if you don't mind reading for like half an hour), you could check out this thread. **https://github.com/flide/8VIM/discussions/65**
If you share your opinion, that discussion might end up less bad for languages with diacritics.

**Reply by Glitchy-Tozier** [reply]
What I thought was really interesting is your opinion on `• layer 2 to 1: 7/10` and `• layer 2 to 6: 7/10`.
Personally, I hate the thought of those movements.

**Reply by ghost** [reply]
> What I thought was really interesting is your opinion on `• layer 2 to 1: 7/10` and `• layer 2 to 6: 7/10`.
> Personally, I hate the thought of those movements.

If you choose your center area big enough the first movement almost results in a figure 8 if you move along the top of the diagonal and the second movement almost results in sort of a circle or ellipse. But I must admit I didn’t put too much thinking into these estimates so feel free to weight these values not too heavy ;)

I’ll have a look at #65. Maybe I’m not good at coding but hopefully I can move this discussion further by adding my 2 cents. I’ve been looking for alternative keyboard layouts for quite some time (Neo2-user since almost 10 years) and this looks like a very interesting project :) I hope I’ll soon be able to use it permanently, too.

**Reply by Glitchy-Tozier** [reply]
Hey, a Neo user!
I recently started learning KOY and i'm still typing in slow-motion. It's getting better though...

**Reply by ghost** [reply]
I took a look at KOY and it looks very interesting but I don’t want to learn a new layout. My job involves quite a lot of typing and I might have some problems if I start a new layout. Anyway our IT specialist will go crazy if I have to ask him to install another layout for me 😆 

My advice for you: Switch all your devices to KOY and force yourself to use it. Print a layout and put it in front of you so you can find the keys you need. The only way to learn a new layout is leaning by doing and typing software didn’t help me at all.
Trust me, it’s worth it 🙂

**Reply by Glitchy-Tozier** [reply]
Ty for the advice! I'm still kind of afraid of forgetting qwertz, so i periodically keep switching between the two.
Also, neo2 is amazing anyways, it might not be worth it switching again if you're comfortable as of now :)

Regarding those movements:
> What I thought was really interesting is your opinion on `• layer 2 to 1: 7/10` and `• layer 2 to 6: 7/10`.

Now that I'm thinking about them, they don't seem as bad as I thought they were in the beginning.

**Comment by davidoskky** [comment]
> 1. How much experience do you have with 8VIM / 8pen?

Not much, I just discovered it and played around for some days. The lack of accents makes it unusable for me.

> 2. If you had 10 "importance-points", how many would you give to fluidity and how many would you give to finger-travel-distance?

6 for fluidity and 4 for travel distance

> 3. On a scale of 1 (worst) to 10 (best), how would you rate those 4 movements? (The points don't have to add up to 10.)

1. 10
2. 10
3. 8
4. 6

> 4. After each of these two Movements, ...

**Movement 1**  

0. 9
1. 5
2. 7
3. 1
4. 1
5. 6
6. 5
7. 10

**Movement 2**

0. 8
1. 0
2. 1
3. 8 
4. 5
5. 10
6. 8
7. 6

**Reply by Glitchy-Tozier** [reply]
Very interesting answers, thank you!

**Comment by Treeoctopus** [comment]
As requested:,

1. How much experience do you have with 8VIM / 8pen?
* a few weeks, learning fast on 8vim. dedicated user of messagease/combokey and have checked in on combokey, minuum, novakey, modality, and any non-qwerty smartphone keyboard that exists. 

2. If you had 10 "importance-points", how many would you give to fluidity and how many would you give to finger-travel-distance?
* 8 to fluidity, 2 to finger travel. I prefer small phones so my answer is skewed in that direction.
3. On a scale of 1 (worst) to 10 (best), how would you rate those 4 movements? (The points don't have to add up to 10.)
* 7, 7, 7, 3 == the biggest path sometimes trips me up but the other ones all seem equally fine to mea

4. After each of these two Movements, ...
* Not sure if I've interpreted this question right

**Comment by kjoetom** [comment]
**1. How much experience do you have with 8VIM / 8pen?**
a few weeks now

**2. If you had 10 "importance-points", how many would you give to fluidity and how many would you give to finger-travel-distance?**
I'd say: 6 to fluidity and 4 to finger-travel-distance

To be on the sure site what you mean in questions 3 and 4, I try to 'translate' the movements based on the original english layout before:
![8Vim-XPad-small](https://user-images.githubusercontent.com/77550697/155242972-824a6515-08cd-47dc-adae-06b00bef632a.png)

**On a scale of 1 (worst) to 10 (best), how would you rate those 4 movements? (The points don't have to add up to 10.)**
o = 10 (quarter-circle)
u = 7 (half-circle)
v = 5 (three-quarter-circle)
w = 3 (full circle)

**4. After each of these two Movements, ...**
0 ... le = 10
1 ... li = 4
2 ... lt = 9
3 ... ls = 2
4 ... ly = 3
5 ... la = 8
6 ... ln = 6
7 ... lo = 7

0 ... ee = 9
1 ... ei = 6
2 ... et = 4
3 ... es = 7
4 ... ey = 5
5 ... ea = 8
6 ... en = 10
7 ... eo = 7

I hope I have understood the movements correctly

**Comment by NK308** [comment]
A notion, that might be important is, that it might be a bit narrow minded to just use mirror and rotational symmetry to extrapolate the results of those questions into every possible transition.
As a left handed person, I find clockwise movement much more comfortable for example.
